aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,218 @@
+/**
+ * @file aesd-circular-buffer.c
+ * @brief Functions and data related to a circular buffer imlementation
+ *
+ * @author Dan Walkes
+ * @date 2020-03-01
+ * @copyright Copyright (c) 2020
+ *
+ */
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#include <linux/slab.h>
+#else
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdio.h>
+#endif
+
+#include "aesd-circular-buffer.h"
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    /**
+    * TODO: implement per description
+    */
+
+    if(buffer == NULL || entry_offset_byte_rtn == NULL)
+        return NULL;
+
+    //If buffer is empty return NULL
+    if((buffer->in_offs ==  buffer->out_offs) && (buffer->full == false))
+    {
+        return NULL;
+    }
+
+    size_t find = char_offset;
+  
+
+    uint8_t length = buffer->out_offs - buffer->in_offs;
+    
+    if(length == 0)
+        if(buffer->full == true)
+            length = AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1;
+    
+  
+
+    //offset shouldn't be greater than total length
+    int tracker = 0;
+    uint8_t off_Seet = buffer->out_offs;
+ 
+    size_t total_size = 0;
+   
+    while((tracker <= length))
+    {
+        total_size += buffer->entry[off_Seet].size;
+        printf("the newsize is  %ld\r\n", total_size);
+        
+        if(off_Seet == (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1))
+            off_Seet= 0;
+        else
+            off_Seet++;
+        printf("offset is %d\r\n", off_Seet);
+        printf("the tracker is %d\r\n", tracker);
+        tracker++;
+    }
+
+    printf("The total size is %ld\r\n", total_size);
+    if(find>total_size)
+        return NULL;
+
+    size_t trace = 0;
+    //size_t size=0;
+    int i= buffer->out_offs;
+    size_t size__ = buffer->entry[buffer->out_offs].size;
+
+    int track_len = 0;
+   int first_iter_flag = 1;
+    printf("\r\nbefore while\r\n");
+    while((track_len <= (length)) && (trace<find) )
```

Comment:
@adna4700 this works but you could simplify by including the check for total length inside this loop and get down to only one loop.